File: editor/en/[494]Target Sum.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        def knapsack(new_target: int) -> int:
            dp = [0] * (new_target + 1)
            dp[0] = 1
            for num in nums:
                for i in range(new_target, num - 1, -1):
                    dp[i] += dp[i - num]
            return dp[new_target]

        total_sum = sum(nums)

        if abs(target) > total_sum or (total_sum + target) % 2 == 1:
            return 0

        return knapsack((total_sum + target) // 2)

# leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[494]Target Sum.java
//leetcode submit region begin(Prohibit modification and deletion)
public class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        int total = Arrays.stream(nums).sum();
        int[] dp = new int[2 * total + 1];
        dp[nums[0] + total] = 1;
        dp[-nums[0] + total] += 1;

        for (int i = 1; i < nums.length; i++) {
            int[] next = new int[2 * total + 1];

            for (int sum = -total; sum <= total; sum++) {
                if (dp[sum + total] > 0) {
                    next[sum + nums[i] + total] += dp[sum + total];
                    next[sum - nums[i] + total] += dp[sum + total];
                }
            }

            dp = next;
        }

        return Math.abs(target) > total ? 0 : dp[target + total];
    }
}
//leetcode submit region end(Prohibit modification and deletion)
